src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,1304 @@
+        symbol_table.push(intrinsic, {intrinsic_loaded_id, SpvStorageClassInput});
+
+        // Annotate that this is the specific builtin
+        auto built_in_kind = starts_with(intrinsic, "Workgroup") ? SpvBuiltInWorkgroupId : SpvBuiltInLocalInvocationId;
+        add_instruction(spir_v_annotations, SpvOpDecorate, {intrinsic_id, SpvDecorationBuiltIn, built_in_kind});
+
+        // Add the builtin to the interface
+        entry_point_interface.push_back(intrinsic_id);
+    }
+
+    // Add the entry point and exection mode
+    add_instruction(spir_v_entrypoints,
+                    SpvOpEntryPoint, entry_point_interface);
+
+    // GLSL-style: each input buffer is a runtime array in a buffer struct
+    // All other params get passed in as a single uniform block
+    // First, need to count scalar parameters to construct the uniform struct
+    std::vector<uint32_t> scalar_types;
+    uint32_t offset = 0;
+    uint32_t param_pack_type_id = next_id++;
+    uint32_t param_pack_ptr_type_id = next_id++;
+    uint32_t param_pack_id = next_id++;
+    scalar_types.push_back(param_pack_type_id);
+    for (const DeviceArgument &arg : args) {
+        if (!arg.is_buffer) {
+            // record the type for later constructing the params struct type
+            scalar_types.push_back(map_type(arg.type));
+
+            // Add a decoration describing the offset
+            add_instruction(spir_v_annotations, SpvOpMemberDecorate, {param_pack_type_id, (uint32_t)(scalar_types.size() - 2), SpvDecorationOffset, offset});
+            offset += arg.type.bytes();
+        }
+    }
+
+    // Add a Block decoration for the parameter pack itself
+    add_instruction(spir_v_annotations, SpvOpDecorate, {param_pack_type_id, SpvDecorationBlock});
+    // We always pass in the parameter pack as the first binding
+    add_instruction(spir_v_annotations, SpvOpDecorate, {param_pack_id, SpvDecorationDescriptorSet, 0});
+    add_instruction(spir_v_annotations, SpvOpDecorate, {param_pack_id, SpvDecorationBinding, 0});
+
+    // Add a struct type for the parameter pack and a pointer to it
+    add_instruction(spir_v_types, SpvOpTypeStruct, scalar_types);
+    add_instruction(spir_v_types, SpvOpTypePointer, {param_pack_ptr_type_id, SpvStorageClassUniform, param_pack_type_id});
+    // Add a variable for the parameter pack
+    add_instruction(spir_v_types, SpvOpVariable, {param_pack_ptr_type_id, param_pack_id, SpvStorageClassUniform});
+
+    uint32_t binding_counter = 1;
+    uint32_t scalar_index = 0;
+    for (const DeviceArgument &arg : args) {
+        uint32_t param_id = next_id++;
+        if (arg.is_buffer) {
+            uint32_t element_type = map_type(arg.type);
+            uint32_t runtime_arr_type = next_id++;
+            uint32_t struct_type = next_id++;
+            uint32_t ptr_struct_type = next_id++;
+            add_instruction(spir_v_types, SpvOpTypeRuntimeArray, {runtime_arr_type, element_type});
```

Comment:
Idle thought: would it be clearer code if you added helper methods like `add_type()`, `add_annotation()` etc which  were shorthand for calling `add_instruction()` with the relevant vector as the first arg?